source/agora/network/Manager.d

```diff
@@ -300,21 +300,51 @@ public class NetworkManager
         in Address address, agora.api.Validator.API api,
         in Hash utxo, in PublicKey key)
     {
-        auto client = new NetworkClient(this.taskman, this.banman,
-            this.config.node.retry_delay, this.config.node.max_retries);
-        if (!client.merge(address, api))
-            assert(0);
+        log.dbg("onHandshakeComplete: {} - (k: {}, utxo: {})", address, key, utxo);
+        this.connection_tasks.remove(address);
 
-        log.dbg("onHandshakeComplete: addresses: {}", client.addresses());
-        client.connections.each!(conn => this.connection_tasks.remove(address));
-        if (this.tryMerge(client))
+        // We have an authenticated client, maybe we already have a client for it
+        if (key !is PublicKey.init)
         {
-            this.required_peers.remove(utxo);
-            return;
+            auto existing_peers = this.peers[].find!(p => p.identity.key == key);
+            if (!existing_peers.empty())
+            {
+                auto prev = existing_peers.front();
+                // There's three possibilities:
+                // a) We don't have an UTXO for this peer and we just got one,
+                //    so record it.
+                // b) We have an UTXO for this peer and got one, don't do anything
+                // c) We have an UTXO for this peer and didn't get one (or got a different one),
+                //    drop the new connection, as it could be another node
+                //    (e.g. backup) not wanting to be known as validator
+                //
+                // Case (c) shows that using keys for identification is tricky,
+                // but we had to do so due to DNS limitations
+                if (prev.identity.utxo !is Hash.init)
+                {
+                    if (utxo != prev.identity.utxo)
+                        return; // Drop it
+                }
+                else if (utxo !is Hash.init)
+                {
+                    prev.setIdentity(utxo, key);
+                    this.required_peers.remove(utxo);
```

Comment:
We can do this at the very beginning (like `conn_tasks.remove()`)and not duplicate in on L339.

Reply:
Actually, L339 means that this will always be the second removal, so I just deleted that line.